chirp/drivers/retevis_ha1g.py

```diff
@@ -0,0 +1,3105 @@
+        self._bank_mappings = []
+        for index, _bank in enumerate(_banks):
+            bank = HA1GBank(self, "%i" % index, "BANK-%i" % index)
+            bank.index = index
+            self._bank_mappings.append(bank)
+
+    def get_num_mappings(self):
+        return len(self._bank_mappings)
+
+    def get_mappings(self):
+        return self._bank_mappings
+
+    def add_memory_to_mapping(self, memory, bank):
+        channels_in_bank = self._channel_numbers_in_bank(bank)
+        if memory.freq == 0:
+            raise errors.RadioError(
+                "Cannot select a channel with empty frequency"
+                )
+        else:
+            ch_num = memory.number
+            if len(channels_in_bank) < 16:
+                if ch_num not in channels_in_bank:
+                    channels_in_bank.add(ch_num)
+                    zone_ch_list = [
+                        x + 2 for x in channels_in_bank
+                    ]
+                    # Adjusting channel numbers
+                    # to match the radio's format
+                    set_zone_ch_list(
+                        self._radio,
+                        bank.index,
+                        sorted(zone_ch_list)
+                        )
+            else:
+                raise Exception("Bank is full, cannot add more channels")
+
+    def remove_memory_from_mapping(self, memory, bank):
+        channels_in_bank = self._channel_numbers_in_bank(bank)
+        ch_num = memory.number
+        if ch_num in channels_in_bank:
+            channels_in_bank.remove(ch_num)
+            zone_ch_list = [x + 2 for x in channels_in_bank]
+            set_zone_ch_list(self._radio, bank.index, sorted(zone_ch_list))
+        else:
+            raise Exception("Memory not found in bank, cannot remove")
+
+    def _channel_numbers_in_bank(self, bank):
+        _zone_data = get_zone_index_list(self._radio)
+        _zone_list = self._radio._memobj.zones
+        if bank.index in _zone_data:
+            _zone = _zone_list[bank.index]
+            _members = [get_ch_index_by_bytes(x) for x in _zone.chindex]
+            return set([int(ch) - 2 for ch in _members if ch != 0xFFFF])
+
+    def get_memory_mappings(self, memory):
+        banks = []
+        for bank in self.get_mappings():
+            if memory.number in self._channel_numbers_in_bank(bank):
+                banks.append(bank)
+        return banks
+
+
+def do_download(self):
+    serial = self.pipe
+    all_bytes = bytearray(self._memsize)
+    handshake_result = handshake(self, serial)
+    if handshake_result == HandShakeStuts.Normal:
+        all_bytes = read_items(self, serial)
+    elif handshake_result == HandShakeStuts.RadioWrong:
+        raise errors.RadioError("Radio not match")
+    elif handshake_result == HandShakeStuts.PwdWrong:
+        raise errors.RadioError("download redio password Wrong")
+    else:
+        raise errors.RadioError("connect Wrong")
+    exit_programming_mode(self)
```

Comment:
You should make this a part of a `finally:` block to make sure it runs even if there is an error. Something like:
```
try:
    all_bytes = read_items(...)
except ...:
    ....
finally:
    exit_programming_mode(self)
```